src/aci/aci_build_F.cc

```diff
@@ -864,4 +862,284 @@ void AdaptiveCI::get_core_excited_determinants(SharedMatrix evecs, DeterminantHa
+        
+            // 4. Screen subspaces
+            double b_sigma = sigma_ * (aci_scale/nbin);
+//            outfile->Printf("\n  Using sigma = %1.5f", b_sigma);
+            double excluded = 0.0;
+            size_t ndet = 0; 
+            for( size_t I = 0, max_I = F_tmp.size(); I < max_I; ++I ){
+                double en = F_tmp[I].first;
+                Determinant& det = F_tmp[I].second;
+                if( excluded + en < b_sigma ){
+                    excluded += en;
+                } else {
+                    F_td.push_back(std::make_pair(en, det));
+                    ndet++;
+                }
+            }
+            outfile->Printf("\n  Added %zu dets from bin %d, (td %d)", ndet, bin, thread_id); 
+            total_excluded += excluded;
+
+        } // End iteration over bins
+
+
+        //6. Merge thread spaces
+        #pragma omp critical
+        {
+            for( auto& pair : F_td ){
+                F_space.push_back(pair);
+            }
+        }
+    } // close threads
+
+    outfile->Printf("\n  Screened out %1.10f Eh of correlation", total_excluded);
+    return total_excluded;
+}
+
+
+
+det_hash<double> AdaptiveCI::get_bin_F_space( int bin, int nbin, SharedMatrix evecs,
+                                                 DeterminantHashVec& P_space) {
+
+    const size_t n_dets = P_space.size();
+    const det_hashvec& dets = P_space.wfn_hash();
+    int nmo = fci_ints_->nmo();
+    std::vector<int> act_mo = mo_space_info_->get_dimension("ACTIVE").blocks();
+    det_hash<double> A_b;
+
+    // Loop over P space determinants
+    for (size_t I = 0; I < n_dets; ++I) {
+        double c_I = evecs->get(I,0);
+        const Determinant& det = dets[I];
+        std::vector<std::vector<int>> noalpha = det.get_asym_occ(nmo,act_mo);
+        std::vector<std::vector<int>> nobeta  = det.get_bsym_occ(nmo,act_mo);
+        std::vector<std::vector<int>> nvalpha = det.get_asym_vir(nmo,act_mo);
+        std::vector<std::vector<int>> nvbeta  = det.get_bsym_vir(nmo,act_mo);
+
+        std::vector<int> aocc = det.get_alfa_occ(nmo);
+        std::vector<int> bocc = det.get_beta_occ(nmo);
+        std::vector<int> avir = det.get_alfa_vir(nmo);
+        std::vector<int> bvir = det.get_beta_vir(nmo);
+
+
+        Determinant new_det(det);
+        // Generate alpha excitations
+        for( int h = 0; h < nirrep_; ++h){
+            for (auto& ii : noalpha[h]) {
+                for (auto& aa : nvalpha[h]) {
+                    new_det = det;
+                    new_det.set_alfa_bit(ii, false);
+                    new_det.set_alfa_bit(aa, true);
+                    size_t hash_val = Determinant::Hash()(new_det);
+                    if ( (hash_val % nbin) == bin) {
+                        double HIJ = fci_ints_->slater_rules_single_alpha(det, ii, aa) * c_I;
+                        if ((std::fabs(HIJ) >= screen_thresh_)) {
+                            if (A_b.count(new_det) == 0) {
```

Comment:
This question is for @jeffschriber  and @tyzhang1993. This operation requires two lookups of the hash table. Is there a way we can get this done in one go? Perhaps something like http://en.cppreference.com/w/cpp/container/map/emplace

Reply:
Yes, it seems that unordered_map initializes values to 0.0, so we can get away with a single A_b[new_det] += HIJ call without any other lookups or logic.